# Sekolah Data Pacmann Final Project: Regional Performance Analysis and Supply Chain Optimization


## A. Dataset Exploration

### A.1. Import Libraries and Load Datasets

In [28]:
import numpy as np
import pandas as pd
import geopandas as gpd
import time

import altair as alt
import seaborn as sns

In [35]:
# Set location of the dataset
filepath = "../datasets/DataCoSupplyChainDataset.csv"
desc = "../datasets/DescriptionDataCoSupplyChain.csv"

# Load datasets, add encoding to prevent error
main_df = pd.read_csv(filepath, encoding='ISO-8859-1')
description = pd.read_csv(desc)

### A.2. Data Understanding and Select Relevant Columns

In [36]:
# Display head, tail, and the df shape
display(main_df.head(3), 
        main_df.tail(3), 
        "Shape: ", main_df.shape)

Type  Days for shipping (real)  Days for shipment (scheduled)  \
0     DEBIT                         3                              4   
1  TRANSFER                         5                              4   
2      CASH                         4                              4   

   Benefit per order  Sales per customer   Delivery Status  \
0          91.250000          314.640015  Advance shipping   
1        -249.089996          311.359985     Late delivery   
2        -247.779999          309.720001  Shipping on time   

   Late_delivery_risk  Category Id   Category Name Customer City  \
0                   0           73  Sporting Goods        Caguas   
1                   1           73  Sporting Goods        Caguas   
2                   0           73  Sporting Goods      San Jose   

  Customer Country Customer Email Customer Fname  Customer Id Customer Lname  \
0      Puerto Rico      XXXXXXXXX          Cally        20755       Holloway   
1      Puerto Rico      XXXXXXXXX          Irene        19492           Luna   
2          EE. UU.      XXXXXXXXX        Gillian        19491      Maldonado   

  Customer Password Customer Segment Customer State           Customer Street  \
0         XXXXXXXXX         Consumer             PR  5365 Noble Nectar Island   
1         XXXXXXXXX         Consumer             PR          2679 Rustic Loop   
2         XXXXXXXXX         Consumer             CA      8510 Round Bear Gate   

   Customer Zipcode  Department Id Department Name   Latitude   Longitude  \
0             725.0              2         Fitness  18.251453  -66.037056   
1             725.0              2         Fitness  18.279451  -66.037064   
2           95125.0              2         Fitness  37.292233 -121.881279   

         Market Order City Order Country  Order Customer Id  \
0  Pacific Asia     Bekasi     Indonesia              20755   
1  Pacific Asia    Bikaner         India              19492   
2  Pacific Asia    Bikaner         India              19491   

  order date (DateOrders)  Order Id  Order Item Cardprod Id  \
0         1/31/2018 22:56     77202                    1360   
1         1/13/2018 12:27     75939                    1360   
2         1/13/2018 12:06     75938                    1360   

   Order Item Discount  Order Item Discount Rate  Order Item Id  \
0            13.110000                      0.04         180517   
1            16.389999                      0.05         179254   
2            18.030001                      0.06         179253   

   Order Item Product Price  Order Item Profit Ratio  Order Item Quantity  \
0                    327.75                     0.29                    1   
1                    327.75                    -0.80                    1   
2                    327.75                    -0.80                    1   

    Sales  Order Item Total  Order Profit Per Order    Order Region  \
0  327.75        314.640015               91.250000  Southeast Asia   
1  327.75        311.359985             -249.089996      South Asia   
2  327.75        309.720001             -247.779999      South Asia   

       Order State Order Status  Order Zipcode  Product Card Id  \
0  Java Occidental     COMPLETE            NaN             1360   
1         Rajastán      PENDING            NaN             1360   
2         Rajastán       CLOSED            NaN             1360   

   Product Category Id  Product Description  \
0                   73                  NaN   
1                   73                  NaN   
2                   73                  NaN   

                                  Product Image  Product Name  Product Price  \
0  http://images.acmesports.sports/Smart+watch   Smart watch          327.75   
1  http://images.acmesports.sports/Smart+watch   Smart watch          327.75   
2  http://images.acmesports.sports/Smart+watch   Smart watch          327.75   

   Product Status shipping date (DateOrders)   Shipping Mode  
0               0             

Type  Days for shipping (real)  Days for shipment (scheduled)  \
180516  TRANSFER                         5                              4   
180517   PAYMENT                         3                              4   
180518   PAYMENT                         4                              4   

        Benefit per order  Sales per customer   Delivery Status  \
180516         141.110001          391.980011     Late delivery   
180517         186.229996          387.980011  Advance shipping   
180518         168.949997          383.980011  Shipping on time   

        Late_delivery_risk  Category Id Category Name Customer City  \
180516                   1           45       Fishing       Bristol   
180517                   0           45       Fishing        Caguas   
180518                   0           45       Fishing        Caguas   

       Customer Country Customer Email Customer Fname  Customer Id  \
180516          EE. UU.      XXXXXXXXX           John          291   
180517      Puerto Rico      XXXXXXXXX           Mary         2813   
180518      Puerto Rico      XXXXXXXXX         Andrea         7547   

       Customer Lname Customer Password Customer Segment Customer State  \
180516          Smith         XXXXXXXXX        Corporate             CT   
180517          Smith         XXXXXXXXX         Consumer             PR   
180518         Ortega         XXXXXXXXX         Consumer             PR   

                   Customer Street  Customer Zipcode  Department Id  \
180516          97 Burning Landing            6010.0              7   
180517  2585 Silent Autumn Landing             725.0              7   
180518           697 Little Meadow             725.0              7   

       Department Name   Latitude  Longitude        Market Order City  \
180516        Fan Shop  41.629959 -72.967155  Pacific Asia   Adelaide   
180517        Fan Shop  18.213350 -66.370575  Pacific Asia   Adelaide   
180518        Fan Shop  18.290380 -66.370613  Pacific Asia  Nagercoil   

       Order Country  Order Customer Id order date (DateOrders)  Order Id  \
180516     Australia                291         1/15/2016 21:00     26024   
180517     Australia               2813         1/15/2016 20:18     26022   
180518         India               7547         1/15/2016 18:54     26018   

        Order Item Cardprod Id  Order Item Discount  Order Item Discount Rate  \
180516                    1004                  8.0                      0.02   
180517                    1004                 12.0                      0.03   
180518                    1004                 16.0                      0.04   

        Order Item Id  Order Item Product Price  Order Item Profit Ratio  \
180516          65129                399.980011                     0.36   
180517          65126                399.980011                     0.48   
180518          65113                399.980011                     0.44   

        Order Item Quantity       Sales  Order Item Total  \
180516                    1  399.980011        391.980011   
180517                    1  399.980011        387.980011   
180518                    1  399.980011        383.980011   

        Order Profit Per Order Order Region        Order State  \
180516              141.110001      Oceania  Australia del Sur   
180517              186.229996      Oceania  Australia del Sur   
180518              168.949997   South Asia         Tamil Nadu   

           Order Status  Order Zipcode  Product Card Id  Product Category Id  \
180516          PENDING            NaN             1004                   45   
180517  PENDING_PAYMENT            NaN             1004                   45   
180518  PENDING_PAYMENT            NaN             1004                   45   

        Product Description  \
180516                  NaN   
180517                  NaN   
180518                  NaN   

                                            Product Image  \
180516  http://images.acmesports.

'Shape: '

(180519, 53)

In [30]:
# Display all column names
display(main_df.columns)

Index(['Type', 'Days for shipping (real)', 'Days for shipment (scheduled)',
       'Benefit per order', 'Sales per customer', 'Delivery Status',
       'Late_delivery_risk', 'Category Id', 'Category Name', 'Customer City',
       'Customer Country', 'Customer Email', 'Customer Fname', 'Customer Id',
       'Customer Lname', 'Customer Password', 'Customer Segment',
       'Customer State', 'Customer Street', 'Customer Zipcode',
       'Department Id', 'Department Name', 'Latitude', 'Longitude', 'Market',
       'Order City', 'Order Country', 'Order Customer Id',
       'order date (DateOrders)', 'Order Id', 'Order Item Cardprod Id',
       'Order Item Discount', 'Order Item Discount Rate', 'Order Item Id',
       'Order Item Product Price', 'Order Item Profit Ratio',
       'Order Item Quantity', 'Sales', 'Order Item Total',
       'Order Profit Per Order', 'Order Region', 'Order State', 'Order Status',
       'Order Zipcode', 'Product Card Id', 'Product Category Id',
       'Product De

In [31]:
# Set options to display all columns
pd.set_option('display.max_columns', None)

# Display datatypes and other information
print(main_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180519 entries, 0 to 180518
Data columns (total 53 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   Type                           180519 non-null  object 
 1   Days for shipping (real)       180519 non-null  int64  
 2   Days for shipment (scheduled)  180519 non-null  int64  
 3   Benefit per order              180519 non-null  float64
 4   Sales per customer             180519 non-null  float64
 5   Delivery Status                180519 non-null  object 
 6   Late_delivery_risk             180519 non-null  int64  
 7   Category Id                    180519 non-null  int64  
 8   Category Name                  180519 non-null  object 
 9   Customer City                  180519 non-null  object 
 10  Customer Country               180519 non-null  object 
 11  Customer Email                 180519 non-null  object 
 12  Customer Fname                

In [34]:
# Display the descriptions for each feature
display(description)

FIELDS  \
0                            Type   
1   Days for shipping (real)        
2   Days for shipment (scheduled)   
3               Benefit per order   
4              Sales per customer   
5                 Delivery Status   
6   Late_delivery_risk              
7                     Category Id   
8                   Category Name   
9                   Customer City   
10               Customer Country   
11                 Customer Email   
12                 Customer Fname   
13                    Customer Id   
14                 Customer Lname   
15              Customer Password   
16               Customer Segment   
17                 Customer State   
18                Customer Street   
19               Customer Zipcode   
20                  Department Id   
21                Department Name   
22                       Latitude   
23                      Longitude   
24                         Market   
25                     Order City   
26                  Order Country   
27              Order Customer Id   
28        order date (DateOrders)   
29                       Order Id   
30         Order Item Cardprod Id   
31            Order Item Discount   
32  Order Item Discount Rate        
33                  Order Item Id   
34  Order Item Product Price        
35        Order Item Profit Ratio   
36            Order Item Quantity   
37                          Sales   
38             Order Item Total     
39         Order Profit Per Order   
40                   Order Region   
41                    Order State   
42                   Order Status   
43                Product Card Id   
44            Product Category Id   
45            Product Description   
46                  Product Image   
47                   Product Name   
48                  Product Price   
49                 Product Status   
50  Shipping date (DateOrders)      
51                  Shipping Mode   

                                          DESCRIPTION  
0                         :  Type of transaction made  
1    :  Actual shipping days of the purchased product  
2   :  Days of scheduled delivery of the purchased...  
3                        :  Earnings per order placed  
4       :  Total sales per customer made per customer  
5   :  Delivery status of orders: Advance shipping...  
6   :  Categorical variable that indicates if send...  
7                            :  Product category code  
8              :  Description of the product category  
9        :  City where the customer made the purchase  
10    :  Country where the customer made the purchase  
11                                :  Customer's email  
12                                   :  Customer name  
13                                     :  Customer ID  
14                               :  Customer lastname  
15                             :  Masked customer key  
16  :  Types of Customers: Consumer , Corporate , ...  
17  :  State to which the store where the purchase...  
18  :  Street to which the store where the purchas...  
19                                :  Customer Zipcode  
20                        :  Department code of store  
21                        :  Department name of store  
22     :  Latitude corresponding to location of store  
23    :  Longitude corresponding to location of store  
24  :  Market to where the order is delivered : Af...  
25                   :  Destination city of the order  
26                :  Destination country of the order  
27                             :  Customer order code  
28                 :  Date on which the order is made  
29                                      :  Order code  
30  :  Product code generated through the RFID reader  
31                       :  Order item discount value  
32                  :  Order item discount percentage  
33                                 :  Order item code  
34              :  Price of products without discount  
35                         :  Order Item Profit R

Select relevant columns

In [15]:
columns_name = [
                'Order Region', 'Order Country', 'Order City', 'Market', 'order date (DateOrders)', 
                'Sales', 'Order Item Quantity', 'shipping date (DateOrders)',
                'Days for shipping (real)', 'Days for shipment (scheduled)', 'Category Name', 
                'Delivery Status', 'Late_delivery_risk', 'Shipping Mode'
                ]

selected_df = main_df[columns_name]

selected_df.head()

Order Region Order Country  Order City        Market  \
0  Southeast Asia     Indonesia      Bekasi  Pacific Asia   
1      South Asia         India     Bikaner  Pacific Asia   
2      South Asia         India     Bikaner  Pacific Asia   
3         Oceania     Australia  Townsville  Pacific Asia   
4         Oceania     Australia  Townsville  Pacific Asia   

  order date (DateOrders)   Sales  Order Item Quantity  \
0         1/31/2018 22:56  327.75                    1   
1         1/13/2018 12:27  327.75                    1   
2         1/13/2018 12:06  327.75                    1   
3         1/13/2018 11:45  327.75                    1   
4         1/13/2018 11:24  327.75                    1   

  shipping date (DateOrders)  Days for shipping (real)  \
0             2/3/2018 22:56                         3   
1            1/18/2018 12:27                         5   
2            1/17/2018 12:06                         4   
3            1/16/2018 11:45                         3   
4            1/15/2018 11:24                         2   

   Days for shipment (scheduled)   Category Name   Delivery Status  \
0                              4  Sporting Goods  Advance shipping   
1                              4  Sporting Goods     Late delivery   
2                              4  Sporting Goods  Shipping on time   
3                              4  Sporting Goods  Advance shipping   
4                              4  Sporting Goods  Advance shipping   

   Late_delivery_risk   Shipping Mode  
0                   0  Standard Class  
1                   1  Standard Class  
2                   0  Standard Class  
3                   0  Standard Class  
4                   0  Standard Class

### A.3. Dataset Summary and Descriptive Statistics

In [39]:
# Display the first few rows
selected_df.head(2)

Order Region Order Country Order City        Market  \
0  Southeast Asia     Indonesia     Bekasi  Pacific Asia   
1      South Asia         India    Bikaner  Pacific Asia   

  order date (DateOrders)   Sales  Order Item Quantity  \
0         1/31/2018 22:56  327.75                    1   
1         1/13/2018 12:27  327.75                    1   

  shipping date (DateOrders)  Days for shipping (real)  \
0             2/3/2018 22:56                         3   
1            1/18/2018 12:27                         5   

   Days for shipment (scheduled)   Category Name   Delivery Status  \
0                              4  Sporting Goods  Advance shipping   
1                              4  Sporting Goods     Late delivery   

   Late_delivery_risk   Shipping Mode  
0                   0  Standard Class  
1                   1  Standard Class

In [42]:
# Display the data types and early missing values detection for each column
selected_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180519 entries, 0 to 180518
Data columns (total 14 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   Order Region                   180519 non-null  object 
 1   Order Country                  180519 non-null  object 
 2   Order City                     180519 non-null  object 
 3   Market                         180519 non-null  object 
 4   order date (DateOrders)        180519 non-null  object 
 5   Sales                          180519 non-null  float64
 6   Order Item Quantity            180519 non-null  int64  
 7   shipping date (DateOrders)     180519 non-null  object 
 8   Days for shipping (real)       180519 non-null  int64  
 9   Days for shipment (scheduled)  180519 non-null  int64  
 10  Category Name                  180519 non-null  object 
 11  Delivery Status                180519 non-null  object 
 12  Late_delivery_risk            

How many regions, countries, and cities (including markets) are included in the dataframe?

In [21]:
print("Region: ", selected_df["Order Region"].nunique(), 
      "\nCountry: ", selected_df["Order Country"].nunique(),
      "\nCity: ",selected_df["Order City"].nunique(),
      "\nMarket: ",selected_df["Market"].nunique())

Region:  23 
Country:  164 
City:  3597 
Market:  5


In [38]:
print("Region: ", selected_df["Order Region"].unique(),
      "\nMarket: ",selected_df["Market"].unique())

Region:  ['Southeast Asia' 'South Asia' 'Oceania' 'Eastern Asia' 'West Asia'
 'West of USA ' 'US Center ' 'West Africa' 'Central Africa' 'North Africa'
 'Western Europe' 'Northern Europe' 'Central America' 'Caribbean'
 'South America' 'East Africa' 'Southern Europe' 'East of USA' 'Canada'
 'Southern Africa' 'Central Asia' 'Eastern Europe' 'South of  USA '] 
Market:  ['Pacific Asia' 'USCA' 'Africa' 'Europe' 'LATAM']


How many categories, delivery statuses, and shipping modes are there in the dataframe?

In [22]:
print("Category Name: ", selected_df["Category Name"].nunique(), 
      "\nDelivery Status: ", selected_df["Delivery Status"].nunique(),
      "\nShipping Mode: ",selected_df["Shipping Mode"].nunique())

Category Name:  50 
Delivery Status:  4 
Shipping Mode:  4


In [23]:
print("\nDelivery Status: ", selected_df["Delivery Status"].unique(),
      "\nShipping Mode: ",selected_df["Shipping Mode"].unique())


Delivery Status:  ['Advance shipping' 'Late delivery' 'Shipping on time' 'Shipping canceled'] 
Shipping Mode:  ['Standard Class' 'First Class' 'Second Class' 'Same Day']


In [44]:
# Display descriptive statistics for numerical columns
selected_df.describe().T

count        mean         std   min  \
Sales                          180519.0  203.772096  132.273077  9.99   
Order Item Quantity            180519.0    2.127638    1.453451  1.00   
Days for shipping (real)       180519.0    3.497654    1.623722  0.00   
Days for shipment (scheduled)  180519.0    2.931847    1.374449  0.00   
Late_delivery_risk             180519.0    0.548291    0.497664  0.00   

                                      25%         50%         75%         max  
Sales                          119.980003  199.919998  299.950012  1999.98999  
Order Item Quantity              1.000000    1.000000    3.000000     5.00000  
Days for shipping (real)         2.000000    3.000000    5.000000     6.00000  
Days for shipment (scheduled)    2.000000    4.000000    4.000000     4.00000  
Late_delivery_risk               0.000000    1.000000    1.000000     1.00000

In [46]:
# Display descriptive statistics for categorical columns
selected_df.describe(exclude=[np.number]).T

count unique               top    freq
Order Region                180519     23   Central America   28341
Order Country               180519    164    Estados Unidos   24840
Order City                  180519   3597     Santo Domingo    2211
Market                      180519      5             LATAM   51594
order date (DateOrders)     180519  65752  12/14/2016 12:29       5
shipping date (DateOrders)  180519  63701     1/5/2016 5:58      10
Category Name               180519     50            Cleats   24551
Delivery Status             180519      4     Late delivery   98977
Shipping Mode               180519      4    Standard Class  107752

<p>Summary of the dataframe:
We can see that the avg `sales` value is around $203.77 with a standard deviation of $132.27 indicating quite a bit of variation in sales amounts. Sales range from a minimum of $9.99 to a maximum of $1,999.99. The avg `order item quantity` is 2.13 items with a minimum of 1 and a maximum of 5 items per order.

When it comes to `days for shipping (real)` on avg it takes about 3.5 days for an order to be shipped (with a standard deviation of 1.62), while the `days for shipment (scheduled)` avg at about 2.93 days. About 54.83% of the orders are at `risk` of being delivered late which is something that might need further investigation later on.

The dataset contains orders from 23 different `regions` with Central America having the highest number of orders (28,341). Orders are placed from 164 `countries` with the United States being the most frequent at 24,840 orders. The orders are spread across 3,597 `cities` with Santo Domingo being the top city with 2,211 orders.

The LATAM `market` has the highest number of orders (51,594) and there are 50 unique `product categories` in the dataset with Cleats being the most popular. Late deliveries account for the majority of the `delivery statuses` (98,977), and the Standard Class `shipping mode` is the most common with 107,752 orders.

Overall, this summary gives us an initial understanding of the dataframe, showing the distribution of sales, order quantities, shipping times, and various categorical variables.</p>

### A.5. Data Visualization and Initial Insights

## B. Data Preparation and Feature Engineering